File: tests/Blokemon.Game.Tests/ResignTests.cs
using Blokemon.Game;
using Shouldly;

namespace Blokemon.Game.Tests;

public sealed class ResignTests
{
    [Test]
    public async Task Resign_IsLegalForBothPlayersInEveryLivePhase()
    {
        var engine = MatchScenario.Engine();
        foreach (var live in LivePhases())
        {
            foreach (var actor in new[] { MatchScenario.FirstPlayer, MatchScenario.SecondPlayer })
            {
                engine
                    .GetLegalActions(live.State, actor)
                    .Count(static action => action.Kind == LegalActionKind.Resign)
                    .ShouldBe(1, $"{live.Name} / {actor.Value}");
            }
        }
    }

    [Test]
    public async Task Resign_CompletesTheMatchForTheOpponentAndClearsPendingRequirements()
    {
        var engine = MatchScenario.Engine();
        foreach (var live in LivePhases())
        {
            foreach (var actor in new[] { MatchScenario.FirstPlayer, MatchScenario.SecondPlayer })
            {
                var applied = (CommandOutcome.Applied)
                    engine.Apply(live.State, ResignCommand(live.State, actor));
                var opponent = live.State.Other(actor);

                applied.State.Winner.ShouldBe(opponent, $"{live.Name} / {actor.Value}");
                applied.State.Phase.ShouldBe(MatchPhase.Complete, live.Name);
                applied.State.PendingEffect.ShouldBeNull(live.Name);
                applied.State.PendingKnockout.ShouldBeNull(live.Name);
                applied.State.PendingBarChits.ShouldBeEmpty(live.Name);
                applied.State.ReplacementPlayer.ShouldBeNull(live.Name);
                applied.State.PendingRoundEnd.ShouldBeFalse(live.Name);
                applied.State.SuddenDeathCount.ShouldBe(live.State.SuddenDeathCount, live.Name);
                applied
                    .Events.Count(matchEvent =>
                        matchEvent.Kind == MatchEventKind.MatchWon && matchEvent.Actor == opponent
                    )
                    .ShouldBe(1, $"{live.Name} / {actor.Value}");
            }
        }
    }

    [Test]
    public async Task Resigning_LeavesNoFurtherLegalActionOrAcceptedCommand()
    {
        var engine = MatchScenario.Engine();
        var state = MatchScenario.BattleState("BLK-003", "BLK-001", ["VIM-BLAZED"], 5);
        var resigned = MatchScenario.Applied(
            engine.Apply(state, ResignCommand(state, MatchScenario.FirstPlayer))
        );

        var endRound = new MatchCommand.EndRound(
            new CommandId("after-resign"),
            resigned.Id,
            MatchScenario.SecondPlayer,
            resigned.Revision
        );
        var rejected = (CommandOutcome.Rejected)engine.Apply(resigned, endRound);

        rejected.Rejection.Code.ShouldBe(CommandRejectionCode.MatchComplete);
        rejected.State.ShouldBe(resigned);
        engine.GetLegalActions(resigned, MatchScenario.FirstPlayer).ShouldBeEmpty();
        engine.GetLegalActions(resigned, MatchScenario.SecondPlayer).ShouldBeEmpty();
    }

    [Test]
    public async Task Resigning_ACompletedMatch_IsRejectedLikeAnyOtherCommand()
    {
        var engine = MatchScenario.Engine();
        var state = MatchScenario.BattleState("BLK-003", "BLK-001", ["VIM-BLAZED"], 5);
        var resigned = MatchScenario.Applied(
            engine.Apply(state, ResignCommand(state, MatchScenario.FirstPlayer))
        );

        var rejected = (CommandOutcome.Rejected)
            engine.Apply(resigned, ResignCommand(resigned, MatchScenario.SecondPlayer));

        rejected.Rejection.Code.ShouldBe(CommandRejectionCode.MatchComplete);
        rejected.State.ShouldBe(resigned);
        rejected.State.Winner.ShouldBe(MatchScenario.SecondPlayer);
    }

    [Test]
    public async Task DeterministicCpu_NeverResignsAndKeepsAnActionInEveryLivePhase()
    {
        var engine = MatchScenario.Engine();
        var cpu = new DeterministicCpu();
        foreach (var live in LivePhases())
        {
            foreach (var actor in new[] { MatchScenario.FirstPlayer, MatchScenario.SecondPlayer })
            {
                var decision = cpu.Choose(engine, live.State, actor);
                var available = engine
                    .GetLegalActions(live.State, actor)
                    .Where(static action => action.Kind != LegalActionKind.Resign)
                    .ToArray();

                decision
                    .Match(
                        static selected => selected.Action.Kind == LegalActionKind.Resign,
                        static _ => false
                    )
                    .ShouldBeFalse($"{live.Name} / {actor.Value}");
                decision
                    .Match(static _ => true, static _ => false)
                    .ShouldBe(available.Length > 0, $"{live.Name} / {actor.Value}");
            }
        }
    }

    private static MatchCommand.Resign ResignCommand(MatchState state, PlayerId actor) =>
        new(new CommandId($"resign:{actor.Value}"), state.Id, actor, state.Revision);

    private static LivePhase[] LivePhases() =>
        [
            new("own turn", MatchScenario.BattleState("BLK-003", "BLK-001", ["VIM-BLAZED"], 5)),
            new("sudden death", SuddenDeathState()),
            new("pending effect choice", PendingEffectChoiceState()),
            new("pending knockout trigger", PendingKnockoutTriggerState()),
            new("pending bar chit trigger", PendingBarChitTriggerState()),
            new("pending replacement", PendingReplacementState()),
        ];

    private static MatchState SuddenDeathState()
    {
        var state = MatchScenario.BattleState("BLK-003", "BLK-001", ["VIM-BLAZED"], 11);
        return state with
        {
            SuddenDeathCount = 1,
            Players = FrozenList<PlayerState>.Create(
                state.Players.Select(static player => player with { BarChitsRemaining = 1 })
            ),
        };
    }

    private static MatchState PendingEffectChoiceState()
    {
        var engine = MatchScenario.Engine();
        var initial = MatchScenario.BattleState("BLK-012", "BLK-001", ["VIM-BLAZED"], 81);
        var bench = MatchScenario.Card(
            "defender-bench",
            "BLK-004",
            MatchScenario.SecondPlayer,
            CardZone.Booth,
            -1
        );
        initial = initial with
        {
            Cards = FrozenList<CardState>.Create(
                initial.Cards.Append(bench).OrderBy(static card => card.Id)
            ),
        };

        var requested = MatchScenario.Applied(
            engine.Apply(initial, MatchScenario.AttackCommand(initial, "BLK-012-B01"))
        );
        requested.Phase.ShouldBe(MatchPhase.AwaitingEffectChoice);
        requested.PendingEffect.ShouldNotBeNull();
        return requested;
    }

    private static MatchState PendingKnockoutTriggerState()
    {
        var engine = MatchScenario.Engine();
        var state = MatchScenario.BattleState(
            "BLK-003",
            "BLK-001",
            ["VIM-BLAZED", "VIM-BLAZED", "VIM-SOBER"],
            103
        );
        var triggerSource = MatchScenario.Card(
            "trigger-source",
            "BLK-026",
            MatchScenario.SecondPlayer,
            CardZone.Booth,
            -1
        );
        var movableVim = MatchScenario.Card(
            "movable-vim",
            "VIM-BEER",
            MatchScenario.SecondPlayer,
            CardZone.Attached,
            -1,
            attachedTo: new CardInstanceId("defender")
        );
        var prize = MatchScenario.Card(
            "prize",
            "VIM-LAIRY",
            MatchScenario.FirstPlayer,
            CardZone.BarChit,
            0
        );
        state = state with
        {
            Cards = FrozenList<CardState>.Create(
                state
                    .Cards.Select(static card =>
                        card.Id.Value == "defender"
                            ? card with
                            {
                                Attachments = FrozenList<CardInstanceId>.Create(
                                    new CardInstanceId("movable-vim")
                                ),
                            }
                            : card
                    )
                    .Append(triggerSource)
                    .Append(movableVim)
                    .Append(prize)
                    .OrderBy(static card => card.Id)
            ),
            Players = FrozenList<PlayerState>.Create(
                state.Players.Select(static player =>
                    player.Id == MatchScenario.FirstPlayer
                        ? player with
                        {
                            BarChitsRemaining = 1,
                        }
                        : player
                )
            ),
        };

        var attacked = MatchScenario.Applied(
            engine.Apply(state, MatchScenario.AttackCommand(state, "BLK-003-B01"))
        );
        attacked.Phase.ShouldBe(MatchPhase.AwaitingTriggerChoice);
        attacked.PendingKnockout.ShouldNotBeNull();
        return attacked;
    }

    private static MatchState PendingBarChitTriggerState()
    {
        var engine = MatchScenario.Engine();
        var state = MatchScenario.BattleState(
            "BLK-003",
            "BLK-001",
            ["VIM-BLAZED", "VIM-BLAZED", "VIM-SOBER"],
            0
        );
        var triggeredPrize = MatchScenario.Card(
            "triggered-prize",
            "BLK-113",
            MatchScenario.FirstPlayer,
            CardZone.BarChit,
            0
        );
        var extraPrize = MatchScenario.Card(
            "extra-prize",
            "VIM-LAIRY",
            MatchScenario.FirstPlayer,
            CardZone.BarChit,
            1
        );
        var defenderBench = MatchScenario.Card(
            "defender-bench",
            "BLK-004",
            MatchScenario.SecondPlayer,
            CardZone.Booth,
            -1
        );
        state = state with
        {
            Cards = FrozenList<CardState>.Create(
                state
                    .Cards.Append(triggeredPrize)
                    .Append(extraPrize)
                    .Append(defenderBench)
                    .OrderBy(static card => card.Id)
            ),
            Players = FrozenList<PlayerState>.Create(
                state.Players.Select(static player =>
                    player.Id == MatchScenario.FirstPlayer
                        ? player with
                        {
                            BarChitsRemaining = 2,
                        }
                        : player
                )
            ),
        };

        var attacked = MatchScenario.Applied(
            engine.Apply(state, MatchScenario.AttackCommand(state, "BLK-003-B01"))
        );
        attacked.Phase.ShouldBe(MatchPhase.AwaitingTriggerChoice);
        attacked.PendingBarChits.ShouldNotBeEmpty();
        return attacked;
    }

    private static MatchState PendingReplacementState()
    {
        var engine = MatchScenario.Engine();
        var state = MatchScenario.BattleState(
            "BLK-026",
            "BLK-151",
            ["VIM-BEER", "VIM-BEER", "VIM-SOBER"],
            97
        );
        var firstPrize = MatchScenario.Card(
            "prize-1",
            "VIM-LAIRY",
            MatchScenario.FirstPlayer,
            CardZone.BarChit,
            0
        );
        var secondPrize = MatchScenario.Card(
            "prize-2",
            "VIM-SOBER",
            MatchScenario.FirstPlayer,
            CardZone.BarChit,
            1
        );
        var defenderBench = MatchScenario.Card(
            "defender-bench",
            "BLK-004",
            MatchScenario.SecondPlayer,
            CardZone.Booth,
            -1
        );
        state = state with
        {
            Cards = FrozenList<CardState>.Create(
                state
                    .Cards.Append(firstPrize)
                    .Append(secondPrize)
                    .Append(defenderBench)
                    .OrderBy(static card => card.Id)
            ),
        };

        var attacked = MatchScenario.Applied(
            engine.Apply(state, MatchScenario.AttackCommand(state, "BLK-026-B01"))
        );
        attacked.Phase.ShouldBe(MatchPhase.AwaitingReplacement);
        attacked.ReplacementPlayer.ShouldBe(MatchScenario.SecondPlayer);
        return attacked;
    }

    private sealed record LivePhase(string Name, MatchState State);
}

File: tests/Blokemon.Game.Tests/ResignTests.fs
namespace Blokemon.Game.Tests

open System.Collections.Immutable
open Blokemon.Game
open Blokemon.Cpu
open FsUnit
open TUnit.Core

[<AutoOpen>]
module private ResignFixtures =

    let resignCommand (state: MatchState) (actor: PlayerId) =
        MatchScenario.Command
            state
            $"resign:{actor.Value}"
            actor
            ImmutableArray<_>.Empty
            MatchAction.Resign

    let private baseState () =
        MatchScenario.BattleState "BLK-039" "BLK-001" [ "VIM-DODGY" ] 11UL

    let private pendingEffectChoiceState () =
        let state = baseState ()
        let command = resignCommand state MatchScenario.FirstPlayer

        { state with
            Phase = MatchPhase.AwaitingEffectChoice
            PendingEffect =
                ValueSome
                    { Command = command
                      Source = CardInstanceId "attacker"
                      Effect = EffectId "BLK-012-B01"
                      Chooser = MatchScenario.SecondPlayer
                      Requirements = ImmutableArray<_>.Empty
                      BeerMatResults = ImmutableArray<_>.Empty
                      AttackStarted = true } }

    let private pendingReplacementState () =
        let state = baseState ()

        let bench =
            MatchScenario.PlainCard "bench" "BLK-004" MatchScenario.SecondPlayer CardZone.Booth 0

        { MatchScenario.WithCards state [ bench ] with
            Phase = MatchPhase.AwaitingReplacement
            ReplacementPlayer = ValueSome MatchScenario.SecondPlayer }

    let livePhases () =
        let suddenDeath =
            { baseState () with
                SuddenDeathCount = 1 }

        [ "own turn", baseState ()
          "sudden death", suddenDeath
          "pending effect choice", pendingEffectChoiceState ()
          "pending replacement", pendingReplacementState () ]

    let bothPlayers = [ MatchScenario.FirstPlayer; MatchScenario.SecondPlayer ]

type ResignTests() =

    [<Test>]
    member _.``resigning should be legal for both players in every live phase``() =
        let engine = MatchScenario.Engine()

        for _, state in livePhases () do
            for actor in bothPlayers do
                engine.GetLegalActions(state, actor)
                |> Seq.filter (fun action -> action.Kind = LegalActionKind.Resign)
                |> Seq.length
                |> should equal 1

    [<Test>]
    member _.``resigning should complete the match for the opponent and abandon every pending requirement``
        ()
        =
        let engine = MatchScenario.Engine()

        for name, state in livePhases () do
            for actor in bothPlayers do
                let applied, events =
                    match engine.Apply(state, resignCommand state actor) with
                    | CommandOutcome.Applied(applied, events) -> applied, events
                    | CommandOutcome.Rejected(_, rejection) ->
                        failwith $"{name}: resigning was rejected with {rejection.Code}."

                let opponent = state.Other actor

                applied.Winner |> should equal (ValueSome opponent)
                applied.Phase |> should equal MatchPhase.Complete
                applied.PendingEffect.IsNone |> should be True
                applied.PendingKnockout.IsNone |> should be True
                applied.PendingBarChits.Length |> should equal 0
                applied.ReplacementPlayer.IsNone |> should be True
                applied.PendingRoundEnd |> should be False
                applied.SuddenDeathCount |> should equal state.SuddenDeathCount

                events
                |> Seq.filter (fun matchEvent ->
                    matchEvent.Kind = MatchEventKind.MatchWon
                    && matchEvent.Actor = ValueSome opponent)
                |> Seq.length
                |> should equal 1

    [<Test>]
    member _.``resigning should leave no further legal action and no acceptable command``() =
        let engine = MatchScenario.Engine()
        let state = MatchScenario.BattleState "BLK-003" "BLK-001" [ "VIM-BLAZED" ] 5UL

        let resigned =
            MatchScenario.Applied(
                engine.Apply(state, resignCommand state MatchScenario.FirstPlayer)
            )

        let endRound =
            MatchScenario.Command
                resigned
                "after-resign"
                MatchScenario.SecondPlayer
                ImmutableArray<_>.Empty
                MatchAction.EndRound

        let rejectedState, rejection =
            MatchScenario.Rejected(engine.Apply(resigned, endRound))

        rejection.Code |> should equal CommandRejectionCode.MatchComplete
        rejectedState |> should equal resigned

        engine.GetLegalActions(resigned, MatchScenario.FirstPlayer).Length
        |> should equal 0

        engine.GetLegalActions(resigned, MatchScenario.SecondPlayer).Length
        |> should equal 0

    [<Test>]
    member _.``resigning a completed match should be refused like any other command``() =
        let engine = MatchScenario.Engine()
        let state = MatchScenario.BattleState "BLK-003" "BLK-001" [ "VIM-BLAZED" ] 5UL

        let resigned =
            MatchScenario.Applied(
                engine.Apply(state, resignCommand state MatchScenario.FirstPlayer)
            )

        let rejectedState, rejection =
            MatchScenario.Rejected(
                engine.Apply(resigned, resignCommand resigned MatchScenario.SecondPlayer)
            )

        rejection.Code |> should equal CommandRejectionCode.MatchComplete
        rejectedState |> should equal resigned
        rejectedState.Winner |> should equal (ValueSome MatchScenario.SecondPlayer)

    [<Test>]
    member _.``the deterministic policy should never resign and should still act in every live phase``
        ()
        =
        let engine = MatchScenario.Engine()
        let cpu = DeterministicCpu()

        for _, state in livePhases () do
            for actor in bothPlayers do
                let decision = cpu.Choose(engine, state, actor)

                let available =
                    engine.GetLegalActions(state, actor)
                    |> Seq.filter (fun action -> action.Kind <> LegalActionKind.Resign)
                    |> Seq.toArray

                let resigned =
                    match decision with
                    | CpuDecision.Selected action -> action.Kind = LegalActionKind.Resign
                    | CpuDecision.NoLegalAction -> false

                let acted =
                    match decision with
                    | CpuDecision.Selected _ -> true
                    | CpuDecision.NoLegalAction -> false

                resigned |> should be False
                acted |> should equal (available.Length > 0)
